kernel/src/log_segment.rs

```diff
@@ -112,13 +121,234 @@ impl LogSegment {
     }
 }
 
+pub(crate) struct LogSegmentBuilder<'a> {
+    fs_client: &'a dyn FileSystemClient,
+    table_root: &'a Url,
+    checkpoint: Option<CheckpointMetadata>,
+    start_version: Option<Version>,
+    end_version: Option<Version>,
+    omit_checkpoint_files: bool,
+    reversed_commit_files: bool,
+}
+impl<'a> LogSegmentBuilder<'a> {
+    pub(crate) fn new(fs_client: &'a dyn FileSystemClient, table_root: &'a Url) -> Self {
+        LogSegmentBuilder {
+            fs_client,
+            table_root,
+            checkpoint: None,
+            start_version: None,
+            end_version: None,
+            omit_checkpoint_files: false,
+            reversed_commit_files: false,
+        }
+    }
+    /// Optionally provide a checkpoint hint that results from reading the `last_checkpoint` file.
+    pub(crate) fn with_checkpoint(mut self, checkpoint: CheckpointMetadata) -> Self {
+        self.checkpoint = Some(checkpoint);
+        self
+    }
+
+    /// Optionally set the start version of the [`LogSegment`]. This ensures that all commit files
+    /// are above this version.
+    #[allow(unused)]
+    pub(crate) fn with_start_version(mut self, version: Version) -> Self {
+        self.start_version = Some(version);
+        self
+    }
+    /// Optionally set the end version of the [`LogSegment`]. This ensures that all commit files
+    /// and checkpoints are below the end version.
+    pub(crate) fn with_end_version(mut self, version: Version) -> Self {
+        self.end_version = Some(version);
+        self
+    }
+    /// Optionally specify that the [`LogSegment`] will not have any checkpoint files. It will only
+    /// be made up of commit files.
+    #[allow(unused)]
+    pub(crate) fn omit_checkpoint_files(mut self) -> Self {
```

Comment:
do we need this? considering we expose the commit files and checkpoint files separately can't the consumer just elect to use one but not the other?

Reply:
Yeah good point.  This may have been an inelegant way of solving this problem: Commit files do not intersect with checkpoints.  Consider the following: 
```
if omit_checkpoint_files {
    checkpoint_files.clear();
}
// ...
if let Some(checkpoint_file) = checkpoint_files.first() {
    commit_files.retain(|log_path| checkpoint_file.version < log_path.version);
}
```
We ensure commit versions and checkpoint versions are disjoint  by `retain`ing commit files that are greater than a checkpoint version. This is the expected behaviour in `Snapshot`. 


Now that I've written and thought about this, I think I can go ahead and do this:
```
    commit_files.retain(|log_path| checkpoint_file.version < log_path.version || log_path.version > start_version);

```

Reply:
For context, we decided to go with `omit_checkpoint_files` in the end. It was a more explicit approach.